technote on federation and saga orchestration (#434)
apollographql/docs: src/content/technotes/TN0040-federation-is-not-a-saga-orchestrator.mdx

---
title: Federation is not a saga orchestrator
id: TN0040
tags: [federation]
---

Apollo Federation is a powerful system for orchestrating queries across a distributed system. For example, you can use Federation to:

- Query data from multiple systems in parallel.
- Sequence calls to multiple systems when there are dependencies between them via [`@key` and `@requires` directives](/federation/entities-advanced).
- Orchestrate any combination of parallel and sequential calls to multiple systems.
- Batch calls to a single system to [solve the infamous N+1 problem](/technotes/TN0019-federation-n-plus-1/).

> - Split a single call into multiple calls with different priorities [using `@defer`](/router/executing-operations/defer-support/).

Reviewer: Cut the periods on these

> Federation will apply the same orchestration capabilities to mutation fields (and their selection sets), but managing state changes across distributes systems typically involves additional requirements. Federation does not provide common functionality such as compensating transactions or data propagation between sibling fields.

Reviewer: ```suggestion
Federation applies these same orchestration capabilities to mutation fields (and their selection sets), but managing state changes across distributed systems typically involves additional requirements. Federation does _not_ provide common functionality for these requirements, such as compensating transactions or data propagation between sibling fields.
```

You will need to implement your own orchestration logic within a single field resolver to implement these requirements. This is due to the design of GraphQL itself.

## Sequencing mutations in GraphQL

The only difference between the root `Query` and `Mutation` types is that [mutation fields are executed serially instead of in parallel](https://spec.graphql.org/October2021/#sec-Normal-and-Serial-Execution). At first glance, this seems like a way to sequence a set of related mutations.

```graphql
mutation DoMultipleThings {
  createAccount {
    success
    account {
      id
    }
  }
  validateAccount {
    success
  }
  setAccountStatus {
    success
  }
}
```

The GraphQL spec ensures that `validateAccount` and `setAccountStatus` will not execute if `createAccount` fails. At first glance, this seems like a way to transactionally execute a set of mutations, but it has some significant downsides:

> - We don't have a way to roll back the changes made by `createAccount` if `validateAccount` or `setAccountStatus` fail.

Reviewer: ```suggestion
- We don't have a way to roll back the changes made by `createAccount` if one of `validateAccount` or `setAccountStatus` fails.
```

```js
const resolvers = {
  Mutation: {
    async createAndValidateAccount() {
      const account = createAccount();

      try {
        validate(account);
        setAccountStatus(account, 'ACTIVE');
      } catch (e) {
        deleteAccount(account);
        return {success: false};
      }

      return {success: true, account};
    }
  }
};
```

By representing this operation in a single field, we can properly handle failure, rollback, and data propagation in one function. We also remove complex error management from our client applications and our schema more clearly expresses the client intent.

## Distributed orchestration

But what if the state changes occur in different services? Because GraphQL and Federation do not provide semantics for distributed transactions, we still need to implement orchestration logic within a single field.

This requirement leads to a few challenging questions:

- What team or domain owns this mutation field?
- In which subgraph service should we implement the resolver?
- How does the resolver communicate with the data services?

> There are no universally correct answers to these questions. It will depend on your organization, your product requirements, and the capabilities of your system.

Reviewer: ```suggestion
There are no universally correct answers to these questions. Your answers will depend on your organization, your product requirements, and the capabilities of your system.
```

A potential strategy is to create one or more subgraphs specifically for orchestrating mutations. They can communicate with data services using REST or RPC. Experience or product teams are often the most appropriate owners of these subgraphs.

> It's tempting to have the orchestrating subgraph resolvers call back to Apollo Router to execute domain-specific mutations. This approach is feasible but requires careful attention to a number of details:

Reviewer 2: interesting - so how does this work in the scenario where you've [secured your subgraphs](https://www.apollographql.com/docs/technotes/TN0021-graph-security/#only-allow-the-router-to-query-subgraphs-directly) where there's only router <-> subgraph communication?

Author: subgraphs should only _receive_ traffic from the router, but they can always _send_ traffic to anything, including the router.

- The calls to domain-specific mutations no longer originate in client applications, so you must propagate client identity and other request metadata.
- GraphOS will view these calls as separate operations.
- If using cloud-native telemetry, you must ensure that Apollo Router receives the trace context to correlate spans.

> - Be wary of circular dependencies. Consider using [Contracts](/graphos/delivery/contracts/) to create variants that expose only the domain-specific mutations and eliminate the potential for loops in an operation.

Reviewer: ```suggestion
- Be wary of circular dependencies. Consider using [contracts](/graphos/delivery/contracts/) to create variants that expose only the domain-specific mutations and eliminate the potential for loops in an operation.
```

---
title: Federation is not a saga orchestrator
id: TN0040
tags: [federation]
---

Apollo Federation is a powerful system for orchestrating queries across a distributed system. For example, you can use federation to:

- Query data from multiple systems in parallel.
- Sequence calls to multiple systems when there are dependencies between them via [`@key` and `@requires` directives](/federation/entities-advanced).
- Orchestrate any combination of parallel and sequential calls to multiple systems.
- Batch calls to a single system to [solve the infamous N+1 problem](/technotes/TN0019-federation-n-plus-1/).
- Split a single call into multiple calls with different priorities [using `@defer`](/router/executing-operations/defer-support/).

> While you can expose mutations in your supergraph, and the query planner will call various systems to resolve mutation fields, Federation does not provide common distributed systems functionality such as compensating transactions, retries, or data propagation.

Reviewer 3: ```suggestion
While you can expose mutations in your supergraph, and the query planner will call various systems to resolve fields returned by a mutation, Federation does not provide common distributed systems functionality such as compensating transactions, retries, or data propagation.
```

```js
const resolvers = {
  Mutation: {
    async createAndValidateAccount() {
      const account = createAccount();

      try {
        validate(account);
      } catch (e) {
        deleteAccount(account);
        return {success: false};
      }

      try {
        setAccountStatus(account, 'ACTIVE');
      } catch (e) {
        deleteAccount(account);
        return {success: false};
      }

      return {success: true, account};
    }
  }
};
```

By representing this operation in a single field, we can properly handle failure, rollback, and data propagation in one function. We also remove complex error management from our client applications and our schema more clearly expresses the client intent.

## Distributed orchestration

But what if the state changes occur in different services? Because GraphQL and Federation do not provide semantics for distributed transactions, we still need to implement orchestration logic within a single field.

This requirement leads to a few challenging questions:

- What team or domain owns this mutation field?
- In which subgraph service should we implement the resolver?
- How does the resolver communicate with the data services?

> There are no right answers to these questions. The answer depends on your organization and the requirements of your system.

Reviewer 3: ```suggestion
There are no universally correct answers to these questions, as the answer depends on your organization and the requirements of your system.
```
